### PR TITLE
More mangling tests

### PR DESCRIPTION
**Pull request by gchatelet**
This allows to test C++ name mangling correctness across platform (Linux, OSX, Windows) and machine architecture (32/64 bits).

Warning : this test needs a valid C++ toolchain (g++ or cl.exe) as well as nm for Linux/OSX or dumpbin.exe for Windows.

### CODE REVIEW

**Comment by yebblies** [comment]
This looks similar to https://gist.github.com/yebblies/a848bd1326b380b471e8 and https://gist.github.com/yebblies/09c20c74c71c3e1fe5ab except it doesn't check the actual calls succeed, only that the mangling matches.

We do need something like this, but I'm not sure it fits well into the test suite.  It is after all not really a runnable test, but an independent test driver.

**Comment by WalterBright** [comment]
I'm not sure what this does. Does it randomly generate C++ code, the D equivalent, and check to see if the names match?

**Comment by gchatelet** [comment]
It's not a fuzzer : it will *not* generate random snippets of code.
The goal is to write C++ snippets and check that the D mangling for it is conform to what the C++ compiler generated. This is needed to improve the test coverage for C++ mangling.

I came to this while rewriting src/cppmangle.c for the third time. One can iterate on one compiler and one platform but we really need to make sure that it runs everywhere and that by improving one we don't break another (eg 32 vs 64).

@yebblies fuzzer is super important to find corner cases but once we find one it should be added to the test suite I believe.

Ultimately I'd like to have a single place to check for correctness of name mangling. Now we have many tests in different files and it's getting more and more confusing.

@WalterBright what would you recommend ?

One thing I can do to improve is to create a subdir with many files. Each file is a template describing how to generate the C++ and D equivalent. Then tests are separated from the test driver as @yebblies said (this would need to be a different test category - not runnable). The DSL could have keywords to repeat the test with all basic types for instance.

What do you think ?

**Comment by gchatelet** [comment]
I came up with this configuration file. It reproduces the tests I have in the pull request (driver is under way). What do you think ?

Strings enclosed with `|` will be replaced by a random name so that the name can be uniquely identified (useful to generate a single C++ file and speedup the test).

If `|TYPES|` is encountered the test is repeated with a predefined set of types ( all basic types + different accessing patterns : ref, pointer, pointer to pointer, const or not ).

`|D_SYMBOL|` is which symbol we want to test for mangling.

`DESCRIPTION` will help debugging in case of failure.

```
// -----------------------------------------------------------------------------
// DESCRIPTION: simple function with no argument.

// CPP:
void |foo|() {}

// D:
void |foo|();

// D_SYMBOL: |foo|

// -----------------------------------------------------------------------------
// DESCRIPTION: simple function with a single basic type argument.

// CPP:
void |foo|(|TYPES|) {}

// D:
void |foo|(|TYPES|);

// D_SYMBOL: |foo|

// -----------------------------------------------------------------------------
// DESCRIPTION: simple function with twice a basic type argument.

// CPP:
void |foo|(|TYPES|, |TYPES|) {}

// D:
void |foo|(|TYPES|, |TYPES|);

// D_SYMBOL: |foo|

// -----------------------------------------------------------------------------
// DESCRIPTION: simple member function with a single basic type argument.

// CPP:
struct |S| {
    void |foo|(|TYPES|);
};

void |S|::|foo|(|TYPES|) {}

// D:
struct |S| {
    void |foo|(|TYPES|);
}

// D_SYMBOL: |S|.|foo|

// -----------------------------------------------------------------------------
// DESCRIPTION: simple const member function with a single basic type argument.

// CPP:
struct |S| {
    void |foo|(|TYPES|) const;
};

void |S|::|foo|(|TYPES|) const {}

// D:
struct |S| {
    void |foo|(|TYPES|) const;
}

// D_SYMBOL: |S|.|foo|

// -----------------------------------------------------------------------------
// DESCRIPTION: simple templated function instanciated with all basic types.

// CPP:
template<typename T>
void |foo|(){};

template void |foo|<|TYPES|>();

// D:
void |foo|(T)();

// D_SYMBOL: |foo|!|TYPES|

// -----------------------------------------------------------------------------
// DESCRIPTION: templated function instanciated with templated type.

// CPP:
struct |S|{};

template<typename T>
void |foo|(){};

// CPP:
template void |foo|<|S|<|S|<|S|<int> > > >();

// D:
struct |S|{}

void |foo|(T)();

// D_SYMBOL: |foo|!(|S|!(|S|!(|S|!int)))

// -----------------------------------------------------------------------------
// DESCRIPTION: templated function with one argument instanciated with int.

// CPP:
template<typename T>
void |foo|(T){};

template void |foo|<int>(int);

// D:
void |foo|(T)(T);

// D_SYMBOL: |foo|!int

// -----------------------------------------------------------------------------
// DESCRIPTION: templated function with three arguments.


// CPP:
template<typename A, typename B, typename C>
void |foo|(A,B&,C){};

template void |foo|<int,char,unsigned>(int,char&,unsigned);

// D:
void |foo|(A,B,C)(A,ref B,C);

// D_SYMBOL: |foo|!(int,char,uint)

// -----------------------------------------------------------------------------
// DESCRIPTION: function with a struct argument in a nested namespace.
// NAMESPACE: a.b.c

// CPP:
struct |S| {};
void |foo|(|S|){}

// D:
struct |S| {}
void |foo|(|S|){}

// D_SYMBOL: |foo|

// -----------------------------------------------------------------------------
// DESCRIPTION: templated function in a templated struct in a nested namespace.
// NAMESPACE: a.b.c

// CPP:
template<typename A>
struct |S| {
    template<typename B> B* |foo|(const B**) const;
};

template<>
template<>
char* |S|<int>::|foo|<char>(const char**) const {
    return nullptr;
};

// D:
struct |S|(A) {
    B* |foo|(B)(const(B)**) const;
}

// D_SYMBOL: |S|!int.|foo|!char

// -----------------------------------------------------------------------------
// DESCRIPTION: templated function in a templated struct in a nested namespace.
// NAMESPACE: a.b.c

// CPP:
struct |S2| {};

template<typename A> struct |S1| {
    template<typename B> B* |foo|(const B**) const;
};

template<>
template<>
|S2|* |S1|<int>::|foo|<|S2|>(const |S2|**) const {
    return nullptr;
};

// D:
struct |S2| {}

struct |S1|(A) {
    B* |foo|(B)(const(B)**) const;
}

// D_SYMBOL: |S1|!int.|foo|!|S2|
```

**Comment by WalterBright** [comment]
My reaction is this is much more complex than necessary. I suggest instead simply coming up with a list of C++ declarations. Compile them, then extract the mangled names. Put the mangled names as strings in a D source file and compare against D generated names.

The thing is, the C++ compilers on various platforms are not going to change the mangling schemes, so it is unnecessary to repeatedly run the C++ compiler on those platforms to generate mangled names. It only has to be done once per platform. If a particular C++ function isn't mangled right by D, then add that function's mangled string to the test suite. Don't need any actual C++ code in it.

Or, just write a bunch of C++ source code, as in:
```
void foo(int, long, double) { }
void foo(long, long, long) { }
```
Write some D code to call those functions, and then check that the program links. If it links, then it's good. I don't understand why a generator is needed (unless it was doing fuzz testing, which is why I asked about randomly generating test cases).

**Comment by yebblies** [comment]
> Write some D code to call those functions, and then check that the program links. If it links, then it's good.

This is enough to check mangling (and my preferred way as it is automatically cross-platform) but to check for ABI correctness we need to actually pass values back and forth and check that the results are correct.

I suggest to:
- Configure this to generate a single giant D file and a single giant C++ file
- Add these files to the test suite set up the same way as the existing C++ interop tests
- Add the generation tool to the tools repo

**Comment by gchatelet** [comment]
So the new produces a [giant d file](https://gist.github.com/gchatelet/2a8ab36c4d18d52cdae6) generated from a giant temporary C++ file.
It is just a tool and will need someone to update the tests on every platform and every architecture upon configuration update.

The code is long but not too complicated. It relies on configuration files which are easy to modify.
Feedback welcome.

File: src/posix.mak [hunk 1]
```diff
@@ -125,6 +126,11 @@ CFLAGS := $(WARNINGS) \
 	-fno-exceptions -fno-rtti \
 	-D__pascal= -DMARS=1 -DTARGET_$(OS_UPCASE)=1 -DDM_TARGET_CPU_$(TARGET_CPU)=1 \
 	$(MODEL_FLAG)
+# GCC Specific
+ifeq ($(HOST_CC), g++)
+CFLAGS := $(CFLAGS) \
+    -std=gnu++98
+endif
```
Comment (ibuclaw) [inline src/posix.mak]: Bad rebase?

Comment (gchatelet) [inline src/posix.mak]: hmm I don't get it. It looks in line with current [HEAD of master](https://github.com/D-Programming-Language/dmd/blob/master/src/posix.mak#L129)

Comment (gchatelet) [inline src/posix.mak]: Ha so I'm not supposed to rebase while doing a pull request I see. Sorry about that I'll revert to previous version of the file.

Comment (ibuclaw) [inline src/posix.mak]: It should be possible. My normal workflow is normally:
checkout -b mybranch
# some work and commits
checkout master
pull upstream master
checkout mybranch
rebase master
push -f origin mybranch

Comment (ibuclaw) [inline src/posix.mak]: The short hand being, don't merge master into your branch, always replay your commits ontop of master.

Comment (gchatelet) [inline src/posix.mak]: I see thx for the explanation :)

**Comment by gchatelet** [comment]
@yebblies I just figured out that I misread your comment. So you're suggesting to not test for `.mangleof` at all and just check if it links ?
At the beginning this will not link because name mangling is broken (at least for templates and STL under Linux and OSX) so my plan with this PR was to give as much information as possible for the person fixing it : the C++ code, the D code and the expected name mangling for the platform.
Now I can keep this generator for me and have this PR to generate C++ and D code as you recommend.

**Comment by yebblies** [comment]
Yes please.  I've found that mangleof tests are more work, because they need to be ported to each mangling scheme, and don't check any other ABI compatibility.  I don't want to just check it links, but that values can be passed back and forth intact.  I've been meaning to take some of the fuzz tester output and turn it into tests like this but haven't gotten around to it.

Checking an actual link has the benefit of flushing out weird things like g++ vs clang compatibility issues, mangling changes in g++ versions, etc

**Comment by ibuclaw** [comment]
This failed to gain any traction.

@gchatelet - if you want to give this another go, please rebase / re-open. :-)
